### PR TITLE
Add PathsAreNotTooLongRule phpstan rule

### LINKED ISSUE #4894: checkout on windows errors with 'filename too long'
checking the current rector revision 51f5a42 I get

```
  HEAD is now at 51f5a4219 [Downgrade] Add ".0" after downgrading the numeric literal separator on floats (#4889)
  error: unable to create file rules/nette-kdyby/tests/Rector/ClassMethod/ReplaceMagicEventPropertySubscriberWithEven
  tClassSubscriberRector/ReplaceMagicEventPropertySubscriberWithEventClassSubscriberRectorTest.php: Filename too long
  error: unable to create file rules/nette/tests/Rector/MethodCall/GetConfigWithDefaultsArgumentToArrayMergeInCompile
  rExtensionRector/GetConfigWithDefaultsArgumentToArrayMergeInCompilerExtensionRectorTest.php: Filename too long
  error: unable to create file rules/solid/tests/Rector/ClassMethod/UseInterfaceOverImplementationInConstructorRector
  /Fixture/skip_class_that_implements_single_interface_but_defines_own_public_methods.php.inc: Filename too long
  error: unable to create file rules/solid/tests/Rector/ClassMethod/UseInterfaceOverImplementationInConstructorRector
  /Fixture/skip_class_that_implements_single_interface_but_uses_trait_with_additional_methods.php.inc: Filename too long
```

see details in https://github.com/rectorphp/rector/issues/4691#issuecomment-745242271

### PR DESCRIPTION
**Pull request by clxmstaab**
closes https://github.com/rectorphp/rector/issues/4894

### CODE REVIEW

File: utils/phpstan-extensions/src/Rule/PathsAreNotTooLongRule.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+{
+    private const MAX_LENGTH = 175;
+    /**
+     * @var string
+     */
+    public const ERROR_MESSAGE = 'The filename "%s" is too long, to be checked out on windows.';
+
+    public function getNodeType(): string
+    {
+        return Classlike::class;
```
Comment (TomasVotruba) [inline utils/phpstan-extensions/src/Rule/PathsAreNotTooLongRule.php:30]: Here could be used special FileNode class

**Comment by clxmstaab** [comment]
I am not sure whether my rule is taken into account or not.

running `composer phpstan` locally gives me 2210.. most similar to
```
 ------------------------------------------------------------------------------------------------------------------------------------------------------
  utils/project-validator/src/Command/ValidateFixtureSuffixCommand.php:16
 ------------------------------------------------------------------------------------------------------------------------------------------------------
  - '#Class namespace "Rector\\Utils\\ProjectValidator\\Command" does not follow PSR\-4 configuration in composer\.json#'
 ------------------------------------------------------------------------------------------------------------------------------------------------------

 ------------------------------------------------------------------------------------------------------------------------------------------------------
  utils/project-validator/src/Command/ValidateSetsCommand.php:23
 ------------------------------------------------------------------------------------------------------------------------------------------------------
  - '#Class namespace "Rector\\Utils\\ProjectValidator\\Command" does not follow PSR\-4 configuration in composer\.json#'
 ------------------------------------------------------------------------------------------------------------------------------------------------------

 ------------------------------------------------------------------------------------------------------------------------------------------------------
  utils/project-validator/src/HttpKernel/SingleConfigKernel.php:10
 ------------------------------------------------------------------------------------------------------------------------------------------------------
  - '#Class namespace "Rector\\Utils\\ProjectValidator\\HttpKernel" does not follow PSR\-4 configuration in composer\.json#'
 ------------------------------------------------------------------------------------------------------------------------------------------------------

 ------------------------------------------------------------------------------------------------------------------------------------------------------
  utils/rule-doc-generator/src/Category/RectorCategoryInferer.php:12
 ------------------------------------------------------------------------------------------------------------------------------------------------------
  - '#Class namespace "Rector\\RuleDocGenerator\\Category" does not follow PSR\-4 configuration in composer\.json#'
 ------------------------------------------------------------------------------------------------------------------------------------------------------
```

I think these are not related to my PR

File: utils/phpstan-extensions/src/Rule/PathsAreNotTooLongRule.php [hunk 2]
```diff
@@ -27,20 +26,16 @@ final class PathsAreNotTooLongRule implements Rule
...
      * @return string[]
      */
     public function processNode(Node $node, Scope $scope): array
     {
-        $fileName = $this->getRealPath($node);
-
-        if (!$fileName) {
-            return [];
-        }
+        $fileName = $scope->getFile();
```
Comment (TomasVotruba) [inline utils/phpstan-extensions/src/Rule/PathsAreNotTooLongRule.php]: :+1:

**Comment by TomasVotruba** [comment]
Unfortunatelly, PHPStan has no way to tell which rule is responsible for the message:

To find out the rule that is repoting error, I search for "does not follow PSR\-4 configuration in composer\.js" in project `/vendor`. There you can find this rule: https://github.com/symplify/symplify/blob/c81a78e25c01e18b2d622bdd1cdc3e017d59cf14/packages/phpstan-rules/src/Rules/CheckClassNamespaceFollowPsr4Rule.php#L23

You can ignore this message for now in `phpstan.neon`. It passes on Linux and CI, so it will be another slash issue. We should not deal with it here.

**Comment by clxmstaab** [comment]
after adding a ignore for `'#.*does not follow PSR.*#'` 

I get 

```
$ vendor/bin/phpstan analyse --ansi --error-format symplify
Note: Using configuration file C:\dvl\GitHub\rector\phpstan.neon.
 2599/2599 [============================] 100%

 ------------------------------------------------------------------------------------------------------------------------------------------------------
  packages/better-php-doc-parser/tests/PhpDocParser/TagValueNodeReprint/TagValueNodeReprintTest.php:56
 ------------------------------------------------------------------------------------------------------------------------------------------------------
  - '#Use value object over return of values#'
 ------------------------------------------------------------------------------------------------------------------------------------------------------



 [ERROR] Found 1 errors
```


still I am not sure my new rule is taken into account. did some print debugging and they don't show up

**Comment by samsonasik** [comment]
@clxmstaab it seems you need to rebase+composer update for phpstan notice.

**Comment by clxmstaab** [comment]
locally seeing the same problem after rebase+compoesr-update

**Comment by samsonasik** [comment]
Please try clear PHPStan cache: 

```bash
vendor/bin/phpstan clear-result-cache
```

https://phpstan.org/user-guide/result-cache#clearing-the-result-cache

File: utils/phpstan-extensions/src/Rule/PathsAreNotTooLongRule.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\PHPStanExtensions\Rule;
+
+use Nette\Utils\Strings;
+use PhpParser\Node;
+use PhpParser\Node\Expr\ClassConstFetch;
+use Rector\Core\PhpParser\Node\CustomNode\FileNode;
```
Comment (staabm) [inline utils/phpstan-extensions/src/Rule/PathsAreNotTooLongRule.php:10]: Should this be `use PHPStan\Node\FileNode;` instead?

File: utils/phpstan-extensions/src/Rule/PathsAreNotTooLongRule.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+use PHPStan\Analyser\Scope;
+use PHPStan\Rules\Rule;
+use Rector\PHPStanExtensions\NodeAnalyzer\SymfonyConfigRectorValueObjectResolver;
+use Rector\PHPStanExtensions\NodeAnalyzer\TypeAndNameAnalyzer;
+use Symplify\SmartFileSystem\SmartFileInfo;
+
+/**
+ * @see \Rector\PHPStanExtensions\Tests\Rule\PathsAreNotTooLongRule\PathsAreNotTooLongRuleTest
+ */
```
Comment (staabm) [inline utils/phpstan-extensions/src/Rule/PathsAreNotTooLongRule.php:19]:
```suggestion
 * @implements Rule<FileNode>
 */
```

File: utils/phpstan-extensions/src/Rule/PathsAreNotTooLongRule.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    /**
+     * @param FileNode $node
+     * @return string[]
+     */
+    public function processNode(Node $node, Scope $scope): array
+    {
+        $fileInfo = $node->getFileInfo();
+        $fileName = $fileInfo->getRealPath();
```
Comment (staabm) [inline utils/phpstan-extensions/src/Rule/PathsAreNotTooLongRule.php:40]:
```suggestion
        $fileName = $scope->getFile();
```

File: utils/phpstan-extensions/tests/Rule/PathsAreNotTooLongRule/PathsAreNotTooLongRuleTest.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+namespace Rector\PHPStanExtensions\Tests\Rule\PathsAreNotTooLongRule;
+
+use Iterator;
+use PHPStan\Rules\Rule;
+use Rector\PHPStanExtensions\Rule\PathsAreNotTooLongRule;
+use Symplify\PHPStanExtensions\Testing\AbstractServiceAwareRuleTestCase;
+use PhpParser\Node\Stmt\ClassLike;
+use Symplify\SmartFileSystem\SmartFileInfo;
+
+final class PathsAreNotTooLongRuleTest extends AbstractServiceAwareRuleTestCase
```
Comment (staabm) [inline utils/phpstan-extensions/tests/Rule/PathsAreNotTooLongRule/PathsAreNotTooLongRuleTest.php:14]: Test could look like https://github.com/phpstan/phpstan-src/commit/ed81c3ad0b9877e6122c79b4afda9d10f3994092#diff-ecaf1e8639b6a502f200830b7f115939ad1ec1e8f9159ad5ed69367092aa8b0dR12 ?

File: utils/phpstan-extensions/src/Rule/PathsAreNotTooLongRule.php (hunk 3 above)
Comment (TomasVotruba) [reply utils/phpstan-extensions/src/Rule/PathsAreNotTooLongRule.php:10]: Yes, it's s PHPStan rule

**Comment by clxmstaab** [comment]
we are getting somewhere.. errors are now showing up.

the issue we try to fix depends on the filename of fixture files though... it seems the phpstan github action only scans the rector-rules for errors...

File: utils/phpstan-extensions/tests/Rule/PathsAreNotTooLongRule/PathsAreNotTooLongRuleTest.php [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        $this->analyse([$filePath], $expectedErrorsWithLines);
+    }
+
+    public function provideData(): Iterator
+    {
+        yield [__DIR__ . '/Fixture/Rector/ShortClassName.php.php', [[]]];
+
+        $longFile = __DIR__ . '/Fixture/Rector/PrettyLongClassNameWaaaaayyyTooLongFileNameWithLoooootsOfCharacters.php.php';
+        $errorMessage = sprintf(PathsAreNotTooLongRule::ERROR_MESSAGE, $longFile, '%s', '%s');
```
Comment (clxmstaab) [inline utils/phpstan-extensions/tests/Rule/PathsAreNotTooLongRule/PathsAreNotTooLongRuleTest.php:31]: not sure how to express a error which is no related to a line

File: utils/phpstan-extensions/tests/Rule/PathsAreNotTooLongRule/PathsAreNotTooLongRuleTest.php [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    public function testRule(string $filePath, array $expectedErrorsWithLines): void
+    {
+        $this->analyse([$filePath], $expectedErrorsWithLines);
+    }
+
+    public function provideData(): Iterator
+    {
+        yield [__DIR__ . '/Fixture/Rector/ShortClassName.php.php', [[]]];
+
+        $longFile = __DIR__ . '/Fixture/Rector/PrettyLongClassNameWaaaaayyyTooLongFileNameWithLoooootsOfCharacters.php.php';
```
Comment (staabm) [inline utils/phpstan-extensions/tests/Rule/PathsAreNotTooLongRule/PathsAreNotTooLongRuleTest.php:30]:
```suggestion
        $longFile = __DIR__ . '/Fixture/Rector/PrettyLongClassNameWaaaaayyyTooLongFileNameWithLoooootsOfCharacters.php';
```

**Comment by TomasVotruba** [comment]
What is missing here? I'd love to merge it so have this check in CI :)

**Comment by staabm** [comment]
> What is missing here? I'd love to merge it so have this check in CI :)

The impl seem to work like expected but the test does not.

**Comment by TomasVotruba** [comment]
Rebased and cleaned up in https://github.com/rectorphp/rector/pull/4998

Thank you!

**Comment by clxmstaab** [comment]
thank you for moving it over the finishing line @TomasVotruba . <3

**Pull request closed without merging**